### DISCUSSION TITLE
Rule engine performances

### DISCUSSION THREAD

**Discussion by MacPaille** [firefly-iii/firefly-iii#5038] [Firefly III] [answered]
Hello @JC5,

Question regarding the performances of the rule engine. I currently have 167 rules which makes my transactions import using the CSV-importer pretty slow. Each transactions takes quite some time before finishing the parsing of all the rules.

I have a lot of small strict rules to handle very specific cases but I also have non-strict ones to set the categories on the transactions. Like for Shopping my rule has 59 triggers either being destination account name is, or starts with or contains... My question is will it be more performant to have then 59 rules with one trigger for these expense accounts or keeping this as one rule is better?

Is there a way to analyze rule parsing performances during an import to see what is possible to optimize?

Thanks!

**Comment by marcquark** [comment]
If the shopping rule handles everything you want to do to shopping transactions in one go you could use the checkbox to completely stop the ruleset when it matches. That way, all the other strict rules won't have to be processed.

Another idea would be to tick the checkboxes next to each trigger (assuming the rule is non-strict) to stop processing further triggers. Though i don't know enough about the inner workings of the rule engine to reliably say whether or not it would actually save you any compute time. Just play around with it i guess?

Maybe you can spin up a toy clone of your current instance and then twist some of the knobs the rule engine has, and throw a reference CSV import at it to see which changes have a notable impact.

**Reply by MacPaille** [reply]
Indeed I think I'll have to use the checkbox to make it stop processing other rules. But for this I'll have to change all my groups to only one or two as the checkbox on the rule itself is to stop processing other rules of the same group. So I suppose it will continue processing all the rules of the other group. My gain would be very minimal in a lot of cases.

For my organization and to avoid having to redo the same stuff multiple times I'm trying to use more generic rules Like first it will set the expense/revenue accounts based on the description (banks are really creative when registering transactions... They find multiple different ways to store a transactions from the same shop haha). Then later I have big ones for each category and part of the triggers are almost all the accounts that need to be set to this category.

Anyway as I'm using mainly the CSV Importer it's not a big issue I let it run the time it needs there is no impact on me. But I like to have things optimized :)

**Comment by JC5** [comment, marked as answer]
> Is there a way to analyze rule parsing performances during an import to see what is possible to optimize?

Not really. Right now, the rule engine is performing pretty well. Each rule is basically a search query into the transaction table. But these are complex queries and there may be a lot of them.

An idea to optimize this is to run the rule engine *after* the import, using the [apply rules](https://docs.firefly-iii.org/firefly-iii/pages-and-features/commandline/#apply-rules) command.

**Reply by MacPaille** [reply]
Don't get me wrong indeed the rule engine is really awesome with so much flexibility! And as I'm using CSV Importer for 99% of my transactions I just leave it run in background while I'm doing something else. I would say the part taking way more time than processing rules is fetching the account list (takes at least 5 minutes).

Indeed the idea to make the rules run AFTER the import of all the transactions may be a big gain of time. Again depending on how it will process everything.
As with the CSV importer you know for which account they have been imported and you could even store the date of the oldest you have imported, it would be possible to trigger each rules/rules groups only for that specific range. Which should improve a LOT the performances than parsing the complete rule set for each transaction.

For example last time I imported my main account I had 164 transactions. Imaging running 164 times the complete ruleset. I feel it's a waste of time and resources :)

But again FF is really and awesome tool!

**Reply by MacPaille** [reply]
I even created bash scripts to parse some on my csv files to reformat them they way I want them to be :)

**Reply by JC5** [reply]
> As with the CSV importer you know for which account they have been imported and you could even store the date of the oldest you have imported, it would be possible to trigger each rules/rules groups only for that specific range. 

This is already the case. It's Firefly III that triggers the rule engine and it [adds a filter](https://github.com/firefly-iii/firefly-iii/blob/main/app/Handlers/Events/StoredGroupEventHandler.php#L75) so only the current transaction is included.

It's true though that it will run x*y times, where x is the number of transactions and y is the number of rules. 😁  And yes, that's kind of wasteful 😮

**Reply by MacPaille** [reply]
At least now I know how it's processed. I'll try to use more the checkbox to stop parsing other rule but as it's limited to the same group will not have a big effect...
It would be nice to also have the same checkbox on the group level saying if a rule did stop the processing of the other rules, stop completely the rules checking as I'm done with what I wanted to do :)

**Reply by JC5** [reply]
There is a check for that in the database, check out `stop_processing` in the `rule_groups` table. I am not sure if it's currently in use.

**Reply by MacPaille** [reply]
> There is a check for that in the database, check out `stop_processing` in the `rule_groups` table. I am not sure if it's currently in use.

If there is no GUI to activate it I don't think it's in use. But that would be nice to have it. Now still have to know how that works if it does what I have explained above or something else

**Reply by JC5** [reply]
Right now, if a rule has "stop_processing", the entire group will stop once a rule from the group has `stop_processing`. I believe this is slightly different from what the UI tells you.

**Comment by JC5** [comment]
I got an alert from a few days ago on a closed issue but for the life of me, I can't find it. But rest assured, CSV account lists will be sorted alphabetically again.

**Reply by MacPaille** [reply]
Haha it was this one: https://github.com/firefly-iii/firefly-iii/issues/4807

Thanks that would help a lot :)
